### PR TITLE
🐛 Only validate shell scripts supported by our parser

### PR DESCRIPTION
**Pull request by chrismcgehee**
* **Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] PR title follows the guidelines defined in  https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Bug fix

* **What is the current behavior?** (You can also link to an open issue here)
The pinned dependency check is failing on some repos because they contain shell scripts that our not supported by the parser we are using. For example, https://github.com/u-boot/u-boot/blob/master/scripts/ld-version.sh is an awk file that we are trying to parse as a shell script. 
Part of #839 

* **What is the new behavior (if this is a feature change)?**
The parser we are using, https://github.com/mvdan/sh, supports parsing the following scripts: bash, sh, and mksh. For this PR, I am only parsing a file if it is using of those 3 shells. If a shebang is present in the file, I base the decision to parse entirely on it, otherwise I use the file extension.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Other information**:
This PR makes it so that some shell scripts (dash, ksh, etc.) will no longer be checked for pinned dependencies. I think this is the most correct thing to do since our parser does not support these shells. If we don't want to lose the checking of these shells, what we could do is attempt to parse these shells, and if parsing fails, swallow the error and continue with the check. This should work ~99% of the time, and in the 1% of times where it doesn't work, the check will run without giving any errors. Let me know if you'd prefer this approach instead.

### CODE REVIEW

File: checks/pinned_dependencies_test.go [hunk 1]
```diff
@@ -515,9 +515,9 @@ func TestGitHubWorflowRunDownload(t *testing.T) {
 			filename: "testdata/github-workflow-wget-across-steps",
 			expected: scut.TestReturn{
 				Errors:        nil,
-				Score:         checker.MinResultScore,
-				NumberOfWarn:  2,
-				NumberOfInfo:  0,
+				Score:         checker.MaxResultScore,
+				NumberOfWarn:  0,
+				NumberOfInfo:  1,
```
Comment (chrismcgehee) [inline checks/pinned_dependencies_test.go:520]: This is because we are no longer parsing `ksh`: https://github.com/ossf/scorecard/blob/8960533b7b38916ad1c0a235643e9468ffbdd152/checks/testdata/github-workflow-wget-across-steps#L43

File: e2e/pinned_dependencies_test.go [hunk 2]
```diff
@@ -49,7 +49,7 @@ var _ = Describe("E2E TEST:"+checks.CheckPinnedDependencies, func() {
 			expected := scut.TestReturn{
 				Errors:        nil,
 				Score:         checker.MinResultScore,
-				NumberOfWarn:  154,
+				NumberOfWarn:  152,
```
Comment (chrismcgehee) [inline e2e/pinned_dependencies_test.go:52]: This is because we are no longer parsing `ksh`: 
https://github.com/ossf-tests/scorecard-check-pinned-dependencies-e2e/blob/91af031ed4e8f5f70d0db2bfd8c126cc27332e1d/.github/workflows/github-workflow-wget-across-steps#L43

**Comment by azeemshaikh38** [comment]
Is this PR ready for review?

**Comment by chrismcgehee** [comment]
> Is this PR ready for review?

Yes. Am I supposed to request reviewers or tag it with something?

**Comment by azeemshaikh38** [comment]
> > Is this PR ready for review?
> 
> Yes. Am I supposed to request reviewers or tag it with something?

You can request reviewers to let them know PR is ready for review. Have added myself and @laurentsimon for now.

File: checks/shell_download_validate_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	"fmt"
+	"io/ioutil"
+	"testing"
+)
+
+func TestIsSupportedShellScriptFile(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		filename string
```
Comment (azeemshaikh38) [inline checks/shell_download_validate_test.go:27]: Nit: add a `name string` field here which can be used to identify the test. Currently we use `filename` to identify the test (line 66).

File: checks/shell_download_validate_test.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt // Re-initializing variable so it is not changed while executing the closure below
+		t.Run(tt.filename, func(t *testing.T) {
+			t.Parallel()
+			var content []byte
+			var err error
+			if tt.filename == "" {
```
Comment (azeemshaikh38) [inline checks/shell_download_validate_test.go:70]: Why do we need to test for empty filename here?

File: checks/shell_download_validate_test.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+				content = make([]byte, 0)
+			} else {
+				content, err = ioutil.ReadFile(tt.filename)
+				if err != nil {
+					panic(fmt.Errorf("cannot read file: %w", err))
```
Comment (azeemshaikh38) [inline checks/shell_download_validate_test.go:75]: Instead of `panic`, use `t.Failf`

File: checks/shell_download_validate_test.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+				}
+			}
+			result := isSupportedShellScriptFile(tt.filename, content)
+			if result != tt.expected {
+				t.Fail()
```
Comment (azeemshaikh38) [inline checks/shell_download_validate_test.go:80]: Add a failure message, showing the `expected` and `got` values. Something like: https://github.com/ossf/scorecard/blob/main/utests/utlib.go#L129

File: checks/shell_download_validate.go [hunk 7]
```diff
@@ -35,6 +35,10 @@ var (
 	shellNames = []string{
 		"sh", "bash", "dash", "ksh", "mksh",
 	}
+	// supportedShells is the list of shells that are supported by mvdan.cc/sh/v3/syntax.
```
Comment (laurentsimon) [inline checks/shell_download_validate.go:38]: you can declare `supportedShells `, and build `shellNames` from it. Take `shellInterpreters` as an example.

File: checks/pinned_dependencies_test.go (hunk 1 above)
Comment (laurentsimon) [reply checks/pinned_dependencies_test.go:520]: instead of ignoring the file, change the shell name in the workflow. We want to make sure this test passes with a supported shell.

File: e2e/pinned_dependencies_test.go (hunk 2 above)
Comment (laurentsimon) [reply e2e/pinned_dependencies_test.go:52]: I've updated the file in https://github.com/ossf-tests/scorecard-check-pinned-dependencies-e2e/blob/main/.github/workflows/github-workflow-wget-across-steps to use `bash` so you need not update this test.

**Review by laurentsimon — APPROVED** [review]

**Review by azeemshaikh38 — APPROVED** [review]

File: checks/shell_download_validate_test.go (hunk 4 above)
Comment (chrismcgehee) [reply checks/shell_download_validate_test.go:70]: Ok, yeah, we don't need that here. I copy-pasted and didn't look to closely at what I copied.

File: checks/shell_download_validate_test.go (hunk 6 above)
Comment (chrismcgehee) [reply checks/shell_download_validate_test.go:80]: Yeah, I was unsure whether I should do that here since the only possibilities are true and false, but I do think it would be helpful.

File: checks/shell_download_validate.go (hunk 7 above)
Comment (chrismcgehee) [reply checks/shell_download_validate.go:38]: Good idea.

**Review by azeemshaikh38 — APPROVED** [review]
